tests/integration/abi_router_test.py

```diff
@@ -0,0 +1,539 @@
+import json
+import re
+from collections import defaultdict
+from dataclasses import asdict
+from pathlib import Path
+
+import pytest
+from graviton.abi_strategy import (
+    ABIArgsMod,
+    RandomABIStrategy,
+    RandomABIStrategyHalfSized,
+)
+from graviton.blackbox import DryRunEncoder
+from graviton.invariant import DryRunProperty as DRProp
+from graviton.invariant import PredicateKind
+
+import pyteal as pt
+from pyteal.compiler.compiler_test import router_app_tester
+from tests.blackbox import (
+    CLEAR_STATE_CALL,
+    ABICallConfigs,
+    Predicates,
+    RouterCallType,
+    RouterSimulation,
+    negate_cc,
+)
+
+NUM_ROUTER_DRYRUNS = 7
+FIXTURES = Path.cwd() / "tests" / "teal" / "router"
+
+
+ALL_ROUTER_CASES, ROUTER_SOURCES = router_app_tester()
+
+ROUTER_CASES, NONTRIV_CLEAR_ROUTER_CASES = ALL_ROUTER_CASES[:-2], ALL_ROUTER_CASES[-2:]
+
+TYPICAL_IAC_OC = pt.MethodConfig(no_op=pt.CallConfig.CALL)
+
+# TEST DRIVERS LEGEND - combines method_configs + predicates
+# * @0 - method: RouterCallType
+#   method == None indicates bare app call
+#
+# * @1 - method_config: MethodConfig
+#   defines how to call the method
+#
+# * @3 - predicates: Predicates ~ dict[DRProp, Any]
+#   these are being asserted after being processed into Invariant's
+#
+# NOTE: the "yacc" routers will simply ignore the case with method `None`
+# as they do not have any bare-app-calls
+QUESTIONABLE_DRIVER: list[tuple[RouterCallType, pt.MethodConfig, Predicates]] = [
+    (
+        "add",
```

Comment:
The `RouterCallType` seems like just a naming used, by passing in `RouterSimulation`, right?

If I am understanding correctly, it is only serving as messaging purpose here?

Reply:
a `RouterCallType` (`str | None | Literal["ClearStateCall"]`) is used drive which method to call. When a `str` it's the method name (admittedly the test can't currently handle `foo()void` AND ALSO `foo()int`), when `None` it's the bare app call, and when `Literal["ClearStateCall"]` it's the clear call. So I'm not sure _messaging_ captures it entirely.

Reply:
ok I see, that correctly described what I understood. Any plan for handling both methods with same name but different args/returns?

Reply:
I see, I tweaked around on my local side, and seems a bit involved, my bad for overlooking the complexity here. 

If you could file a issue on this one, I would love to take that for improvement. But current one is sufficiently good (so long as we do not have method name duplicated).

Reply:
I've created #662

Reply:
thanks, resolving.